server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
         return indices.get(indices.size() - 1);
     }
 
+    public Index getWriteIndex(DocWriteRequest<?> request, Metadata metadata) {
+        if (type != Type.TSDB) {
+            return getWriteIndex();
+        }
+
+        if (request instanceof IndexRequest == false || request.opType() != DocWriteRequest.OpType.CREATE) {
+            return getWriteIndex();
+        }
+
+        // TODO: this really needs be parsed in a streaming manner:
+        String timestampAsString = (String) ((IndexRequest) request).sourceAsMap().get("@timestamp");
+        long timestamp = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(timestampAsString);
```

Comment:
We should probably throw an error in the mapping if the field doesn't have the default formatter.

Hmmmmmm.......... What about nanos?

Reply:
I think we should check mapping in the cluster state and try to detect what formatter to use here.
This should also allow this code to parse nanos. I will add an TODO for this.